File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.ts
function findSubstring(s: string, words: string[]): number[] {
  const total = words.length;
  const chunk = words[0].length;
  const count = new Map<string, number>();
  const found = new Map<string, number>();
  const ans: number[] = [];
  for (const word of words) count.set(word, (count.get(word) ?? 0) + 1)
  for (let word: string, val: number, i = 0, m = s.length - total * chunk; i <= m; i++) {
    found.clear();
    let added = 0;
    for (let j = i, m = i + total * chunk - chunk; j <= m; j += chunk) {
      word = s.substr(j, chunk);
      val = (found.get(word) ?? 0) + 1;
      if (val > (count.get(word) ?? 0)) break;
      found.set(word, val);
      added++;
    }
    if (added === total) ans.push(i);
  }
  return ans;
};

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.go
func findSubstring(s string, words []string) []int {
  total := len(words)
  chunk := len(words[0])
  count := map[string]int{}
  found := map[string]int{}
  ans := []int{}
  for _, word := range words {
    count[word]++
  }
  for i, m := 0, len(s) - total * chunk; i <= m; i++ {
    if count[s[i: i + chunk]] == 0 { continue }
    found = map[string]int{}
    added := 0
    for j, m := i, i + total * chunk - chunk; j <= m; j += chunk {
      word := s[j: j + chunk]
      if found[word] + 1 > count[word] { break }
      found[word]++
      added++
    }
    if added == total {
      ans = append(ans, i)
    }
  }
  return ans
}